accord-core/src/main/java/accord/coordinate/CheckOn.java

```diff
@@ -94,23 +94,39 @@ protected AbstractRoute route()
     @Override
     protected boolean isSufficient(Id from, CheckStatusOk ok)
     {
-        return ((CheckStatusOkFull)ok).sufficientFor(route()).compareTo(sufficient) >= 0;
+        KeyRanges rangesForNode = topologies().computeRangesForNode(from);
+        PartialRoute scope = this.route.slice(rangesForNode);
+        return isSufficient(scope, ok);
     }
 
     @Override
-    protected void onDone(Done done, Throwable failure)
+    protected boolean isSufficient(CheckStatusOk ok)
     {
+        return isSufficient(route, ok);
+    }
+
+    protected boolean isSufficient(AbstractRoute scope, CheckStatusOk ok)
+    {
+        return ((CheckStatusOkFull)ok).sufficientFor(scope).compareTo(sufficient) >= 0;
+    }
+
+    @Override
+    protected void onDone(Success success, Throwable failure)
+    {
+        Preconditions.checkState((success == null) != (failure == null));
         if (failure != null)
         {
             callback.accept(null, failure);
         }
-        else if (merged == null || merged.saveStatus == NotWitnessed)
-        {
-            callback.accept(CheckStatusOkFull.NOT_WITNESSED, null);
-        }
         else
         {
-            new OnDone().start();
+            if (success == Success.Success)
```

Comment:
This wouldn't also be true with quorum?

Reply:
No, if anything the opposite. The point of `isSufficient` is to decide when we hit `Accept`. If we hit `AcceptQuorum` we should be `!isSufficient` here.

Reply:
Oh right... AcceptQuorum is bubbling up here, but it's actually an indication of insufficient. I think that enum value could be better named.

Should it be `Insufficient` and `Sufficient` instead of `Accept` and `AcceptQuorum`?

Reply:
Perhaps `AcceptNow` and `AcceptQuorum`? It feels like the concept of sufficiency is fairly local, and probably shouldn't bubble up to the tracker - it only really knows if we're done _now_ or done when we hit a quorum. But perhaps we can standardise terminology between the two.

Reply:
Perhaps simply `Success` and `SuccessIfQuorum`?

Reply:
I had a look at a thesaurus, and another candidate is `Approve` and `ApproveIfQuorum`. I'll go with that for now.

Reply:
Thanks that feels more intuitive to me.